release.py

```diff
@@ -0,0 +1,289 @@
+#! /usr/bin/env python
+# Tahoe-LAFS -- secure, distributed storage grid
+#
+# Copyright © 2006-2012 The Tahoe-LAFS Software Foundation
+#
+# This file is part of Tahoe-LAFS.
+#
+# See the docs/about.rst file for licensing information.
+import os
+import shutil
+import re
+import sys
+import subprocess
+import argparse
+import datetime
+import pathlib
+
+
+class bcolors:
+    HEADER = "\033[95m"
+    OKBLUE = "\033[94m"
+    OKCYAN = "\033[96m"
+    OKGREEN = "\033[92m"
+    WARNING = "\033[93m"
+    FAIL = "\033[91m"
+    ENDC = "\033[0m"
+    BOLD = "\033[1m"
+    UNDERLINE = "\033[4m"
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--clean", action="store_true", help="Cancel existing release process, clean files"
+)
+parser.add_argument(
+    "--ignore-deps", action="store_true", help="Ignore dependency checks"
+)
+parser.add_argument("--fin", action="store_true", help="Finish release proccess")
+parser.add_argument("--sign", type=str, help="Signing key")
+parser.add_argument("--ticket", type=int, help="Ticket number", required=True)
+parser.add_argument(
+    "--repo",
+    type=str,
+    help="Destination repo (fork), example: git@github.com:meejah/tahoe-lafs.git",
+)
+parser.add_argument("--retry", action="store_true", help="Retry release")
+parser.add_argument("--tag", type=str, help="Release tag", required=True)
+args = parser.parse_args()
+
+
+TAG = args.tag
+TICKET = args.ticket
+TODAY = datetime.date.today()
+BRANCH = "{ticket}.release-{tag}".format(
+    ticket=TICKET, tag=TAG
+)  # looks like XXXX.release-1.16.0
+RELEASE_TITLE = "Release {tag} ({date})".format(tag=TAG, date=TODAY)
+RELEASE_FOLDER = pathlib.Path("../tahoe-release-{0}".format(TAG))
+RELEASE_PROGRESS = pathlib.Path("../.tahoe-release-{0}-progress".format(TAG))
```

Comment:
This would be better as a file inside `RELEASE_FOLDER` I think, to keep things together. Maybe `RELEASE_FOLDER.join(".release-progress")` or similar?

Reply:
Not sure if I would have to specify that, it should not be bundled with the release files but then, the progress folder is deleted just like the release folder at the end.

But most importantly, the first step is cloning the repo which create the "RELEASE_FOLDER" so the tracking of progress predates the existence of the RELEASE_FOLDER.

Reply:
the first time you run `record_step` is after the clone, so it'll exist.

Reply:
Why is that a problem? It just checks if the file exists (and in either case it won't).

The reason I don't like it is you have to remember to delete two things if/when things go sideways and it's in a not-very-expected location (neither the current checkout/repo nor the newly-created one).

Reply:
It's a problem because it tries to check if the clone step was already done before the clone step is done! (The first time), and the whole progress thing is there so that if the clone step succeeds and something else fails the script does not need to clone again.

So it can't check if the clone step has worked the first time without first cloning and the progress won't exist.

Secondly, no one has to remember to delete anything. `clean()` is called at the end of the script and can bed forced by passing `--clean` flag.

Reply:
Right, so if the clone _isn't_ done then the file doesn't exist (whether or not the directory the file might be in exists or not doesn't matter). This is true no matter where that "clone is done" file exists. Keeping all the information together makes more sense to me.

Reply:
```
if not os.path.exists(RELEASE_PROGRESS):
    os.mkdir(RELEASE_PROGRESS)

start_release()
```

Reply:
Except remembering to delete : Which is properly address, what other problem is there with this current arrangement?

Reply:
I don't feel it has been addressed, and it was specifically about "when there are problems" (e.g. script hasn't finished etc). Sure, you could remember to run `--clean` but I just find it weird to have the status information elsewhere when it seems straightforward to put it alongside the clean checkout.

I'm not interested in debating this any further, if you don't want to change it I will next time I do a release.